### PR TITLE
DX: remove `symfony/phpunit-bridge`

### PR DESCRIPTION
**Pull request by kubawerlos**
Both [listeners](https://github.com/sebastianbergmann/phpunit/issues/3389) and [`expectDeprecation`](https://github.com/sebastianbergmann/phpunit/commit/a2c784c39bf5741caf436f92fc8c5bfb91bac025) got removed in PHPUnit 10, so to keep checking for deprecations we need to have our implementation.

We can move to the event system that got [introduced in PHPUnit 10](https://github.com/sebastianbergmann/phpunit/blob/10.0.0/ChangeLog-10.0.md) when we use PHPUnit 10+ (requires supporting PHP 8.1+ only).

### CODE REVIEW

File: tests/TestCase.php [hunk 1]
```diff
@@ -24,7 +23,25 @@
...
+
+    /** @var array<string> */
+    private array $expectedDeprecations = [];
+
+    /** @var array<string> */
+    private array $actualDeprecations = [];
+
+    protected function tearDown(): void
+    {
+        if (null !== $this->previouslyDefinedErrorHandler) {
```
Comment (keradus) [inline tests/TestCase.php:37]: is anything around https://github.com/sebastianbergmann/phpunit/issues/5532 helping us to avoid local implementation?

Comment (kubawerlos) [reply tests/TestCase.php:37]: I wasn't able to figure anything out that would still work with PHP 7.4 and PHPUnit 9.

File: tests/FileRemovalTest.php [hunk 2]
```diff
@@ -127,6 +127,8 @@ public function testWakeup(): void
      *
      * @runInSeparateProcess
      *
+     * @preserveGlobalState disabled
```
Comment (keradus) [inline tests/FileRemovalTest.php:130]: intrested that it was not needed so far

File: tests/TestCase.php [hunk 3]
```diff
@@ -24,7 +23,25 @@
  */
 abstract class TestCase extends BaseTestCase
 {
-    use ExpectDeprecationTrait;
+    /** @var null|callable */
+    private $previouslyDefinedErrorHandler;
+
+    /** @var array<string> */
```
Comment (keradus) [inline tests/TestCase.php:29]:
```suggestion
    /** @var list<string> */
```

if indexes are 0..n and not `1, 5, 10`, I want to use `list` for whole project

File: tests/TestCase.php [hunk 4]
```diff
@@ -24,7 +23,25 @@
...
+    private array $expectedDeprecations = [];
+
+    /** @var array<string> */
```
Comment (keradus) [inline tests/TestCase.php:32]:
```suggestion
    /** @var list<string> */
```

File: tests/TestCase.php [hunk 5]
```diff
@@ -24,7 +23,25 @@
...
+    private array $actualDeprecations = [];
+
+    protected function tearDown(): void
+    {
+        if (null !== $this->previouslyDefinedErrorHandler) {
+            foreach ($this->expectedDeprecations as $expectedDeprecation) {
+                self::assertContains($expectedDeprecation, $this->actualDeprecations);
+            }
```
Comment (keradus) [inline tests/TestCase.php:40]: what if `count(actual) > count(expected)` ?
we could miss some exceptions. I suggest to check them too

File: tests/TestCase.php [hunk 6]
```diff
@@ -35,4 +52,27 @@ final public function testNotDefiningConstructor(): void
             $reflection->getName(),
         );
     }
+
+    /**
+     * @TODO change access to protected and pass the parameter when PHPUnit 9 support is dropped
+     */
```
Comment (keradus) [inline tests/TestCase.php:58]:
```suggestion
     * @param string $message
     */
```

possible ?

**Review by keradus — APPROVED** [review]
overall OK, left few minor comments

File: tests/FileRemovalTest.php (hunk 2 above)
Comment (kubawerlos) [reply tests/FileRemovalTest.php:130]: I wouldn't call it "interesting", more like hidden "gem" of PHPUnit, I've extracted it in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7579#issuecomment-1864910163

File: tests/TestCase.php (hunk 5 above)
Comment (kubawerlos) [reply tests/TestCase.php:40]: I will so that in separate PR, will look quite funny, I promise.

File: tests/TestCase.php (hunk 6 above)
Comment (kubawerlos) [reply tests/TestCase.php:58]: Ondrej says "no":
```
 ------ ----------------------------------------------------------
  Line   tests/TestCase.php
 ------ ----------------------------------------------------------
  60     PHPDoc tag @param references unknown parameter: $message
 ------ ----------------------------------------------------------
```

Comment (keradus) [reply tests/TestCase.php:58]: yeah, was afraid some sca will fail on "future param"

**Pull request merged**